org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/DirtyStateManager.java

```diff
@@ -35,24 +36,30 @@
 public class DirtyStateManager extends AbstractResourceDescriptionChangeEventSource implements IDirtyStateManager, IDirtyStateManagerExtension {
 
 	private ConcurrentMap<URI, IDirtyResource> managedResources;
+	private ConcurrentMap<URI, IDirtyResource> discardCandidates;
```

Comment:
Introducing an additional field will require us to work with synchronization to keep both states consistent.